File: LeetCode/918. Maximum Sum Circular Subarray.py
# https://leetcode.com/problems/maximum-sum-circular-subarray/description/

from test import test


class Solution:
    def maxSubarraySumCircular(self, nums: list[int]) -> int:
        local_max = nums[0]
        global_max = nums[0]

        local_min = nums[0]
        global_min = nums[0]

        total_sum = nums[0]

        for num in nums[1:]:
            local_max = max(local_max + num, num)
            global_max = max(global_max, local_max)

            local_min = min(local_min + num, num)
            global_min = min(global_min, local_min)

            total_sum += num

        if global_max < 0:
            return global_max

        return max(global_max, total_sum - global_min)


solution = Solution()


test(
    solution.maxSubarraySumCircular,
    [
        {
            "input": [[1, -2, 3, -2]],
            "expected": 3,  # 3
        },
        {
            "input": [[5, -3, 5]],
            "expected": 10,  # 7
        },
        {
            "input": [[-3, -2, -3]],
            "expected": -2,  # -2
        },
    ],
)

File: LeetCode/918. Maximum Sum Circular Subarray.js
// https://leetcode.com/problems/maximum-sum-circular-subarray/description/

const { test } = require('../test');

const maxSubarraySumCircular = (nums) => {
  let totalSum = nums[0];

  let localMax = nums[0];
  let globalMax = nums[0];

  let localMin = nums[0];
  let globalMin = nums[0];

  for (let i = 1; i < nums.length; i++) {
    localMax = Math.max(localMax + nums[i], nums[i]);
    globalMax = Math.max(globalMax, localMax);

    localMin = Math.min(localMin + nums[i], nums[i]);
    globalMin = Math.min(globalMin, localMin);

    totalSum += nums[i];
  }

  if (globalMax < 0) {
    return globalMax;
  }

  return Math.max(globalMax, totalSum - globalMin);
};

test(maxSubarraySumCircular, [
  {
    input: [[1, -2, 3, -2]],
    expected: 3, // 3
  },
  {
    input: [[5, -3, 5]],
    expected: 10, // 7
  },
  {
    input: [[-3, -2, -3]],
    expected: -2, // -2
  },
]);
